### PR TITLE
HIVE-27654: Iceberg: Cherry-Pick commit to a branch.

### PR DESCRIPTION
**Pull request by ayushtkn**
### What changes were proposed in this pull request?

Allow cherry-pick operation

### Why are the changes needed?

To enable cherry-pick for iceberg tables

### Does this PR introduce _any_ user-facing change?

They can cherry-pick a commit

### Is the change a dependency upgrade?

No

### How was this patch tested?

UT

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java [hunk 1]
```diff
@@ -124,6 +126,11 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
       AlterTableExecuteSpec spec =
           new AlterTableExecuteSpec(FAST_FORWARD, new FastForwardSpec(branchName, targetBranchName));
       desc = new AlterTableExecuteDesc(tableName, partitionSpec, spec);
+    } else if (HiveParser.KW_CHERRY_PICK == executeCommandType.getType()) {
+      ASTNode child = (ASTNode) command.getChild(1);
+      long snapshotId = Long.parseLong(child.getText());
+      AlterTableExecuteSpec spec = new AlterTableExecuteSpec(CHERRY_PICK, new CherryPickSpec(snapshotId));
+      desc = new AlterTableExecuteDesc(tableName, partitionSpec, spec);
```
Comment (zhangbutao) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java:133]: I see each `else if` clause has partial same code snippet. e.g.
`ASTNode child = (ASTNode) command.getChild(1);`
and
` desc = new AlterTableExecuteDesc(tableName, partitionSpec, spec);`

Could we extract these common logic to make code more cleaner?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 2]
```diff
@@ -1400,7 +1405,7 @@ private static boolean hasParquetNestedTypeWithinListOrMap(Properties tableProps
    * Generates {@link JobContext}s for the OutputCommitter for the specific table.
    * @param configuration The configuration used for as a base of the JobConf
    * @param tableName The name of the table we are planning to commit
-   * @param overwrite If we have to overwrite the existing table or just add the new data
+   * @param branchName If we have to overwrite the existing table or just add the new data
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:1408]: Need to change the description of `branchName`.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    // make 2 new inserts to the main branch
+    shell.executeStatement(String.format("INSERT INTO TABLE %s VALUES(5),(6)", identifier.name()));
+    shell.executeStatement(String.format("INSERT INTO TABLE %s VALUES(7),(8)", identifier.name()));
+
+    icebergTable.refresh();
+    long id2 = icebergTable.currentSnapshot().snapshotId();
+
+    Assert.assertNotEquals(id1, id2);
+
+    // Rollback to the previous id before inserts.
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java:76]: This comment seems to be unclear. I only see `rollback `and do not see `insert `after `rollback`.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java (hunk 1 above)
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java]: I have refactored the code, give a check if it looks better. happy to modify even further if you have suggestions :-)

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 2 above)
Comment (ayushtkn) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: done

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java (hunk 3 above)
Comment (ayushtkn) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java]: I made the comment a bit more clear, was talking about the inserts done above, like rollback the table to the state before the above inserts were done

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java (hunk 1 above)
Comment (zhangbutao) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java]: IMO it's good enough. It's more clearly than before. Thx.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 2 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: Now i see. Thx.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java (hunk 3 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergCherryPick.java]: Now i see. Thx.

**Review by zhangbutao — APPROVED** [review]
LGTM +1

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java [hunk 4]
```diff
@@ -73,59 +75,85 @@ protected void analyzeCommand(TableName tableName, Map<String, String> partition
...
-        ZoneId timeZone = SessionState.get() == null ? new HiveConf().getLocalTimeZone() : SessionState.get().getConf()
-            .getLocalTimeZone();
-        TimestampTZ time = TimestampTZUtil.parse(PlanUtils.stripQuotes(child.getText()), timeZone);
-        spec = new AlterTableExecuteSpec(ROLLBACK, new RollbackSpec(TIME, time.toEpochMilli()));
-      } else {
-        spec = new AlterTableExecuteSpec(ROLLBACK, new RollbackSpec(VERSION,
-            Long.valueOf(child.getText())));
-      }
-      desc = new AlterTableExecuteDesc(tableName, partitionSpec, spec);
+      desc = getRollbackDesc(tableName, partitionSpec, (ASTNode) command.getChild(1));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/execute/AlterTableExecuteAnalyzer.java:78]: could we use `switch` instead on if-else?

**Review by deniskuzZ — COMMENTED** [review]
LGTM, 1 minor comment

**Review by deniskuzZ — APPROVED** [review]
+1, pending tests

**Pull request merged**